sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1028,15 +1028,19 @@ _pi_queue::resetCommandList(pi_command_list_ptr_t CommandList,
     EventList.clear();
   } else {
     // For immediate commandlist reset only those events that have signalled.
+    std::vector<pi_event> UnsetEventsList{};
     for (auto it = EventList.begin(); it != EventList.end(); it++) {
       std::scoped_lock<pi_shared_mutex> EventLock((*it)->Mutex);
       ze_result_t ZeResult =
           ZE_CALL_NOCHECK(zeEventQueryStatus, ((*it)->ZeEvent));
       if (ZeResult == ZE_RESULT_SUCCESS) {
-        std::move(it, it, std::back_inserter(EventListToCleanup));
-        EventList.erase(it, it);
+        EventListToCleanup.push_back(std::move((*it)));
+      } else {
+        UnsetEventsList.push_back(std::move((*it)));
       }
     }
+    EventList.clear();
+    EventList = std::move(UnsetEventsList);
```

Comment:
seems like some redundant copying, why do you need a new `UnsetEventsList` at all?

Reply:
Trying to delete elements from EventList inside a loop which is iterating over EventList does not work. Odd things happen, like skipping elements in the original EventList. That caused the leak. It is safer to iterate over EventList, gather remaining elements in another list and then update EventList.

Reply:
Or maybe `std::remove_if(begin, end, predicate/lambda)` ?

Reply:
A combination of the suggestions was used. The additional list has been eliminated and deletions from EventList are done in-place.